File: 9주차/이라은/무인도여행.py
from collections import deque

def solution(maps):
    answer = []
    graph = []
    
    for i in range(len(maps)):
        graph.append(list(maps[i]))

    visited = [[False] * len(graph[0]) for _ in range(len(graph))]
    q = deque()
    cnt_lst = []
    
    dx = [0,0,1,-1]
    dy = [1,-1,0,0]
    
    def bfs(x,y,cnt):
        q.append((x,y))
        cnt += int(graph[x][y])
        while q:
            x, y = q.popleft()
            for i in range(4):
                nx = x + dx[i]
                ny = y + dy[i]
                if 0 <= nx < len(graph) and 0 <= ny < len(graph[0]) and not visited[nx][ny] and graph[nx][ny] != 'X':
                    visited[nx][ny] = True
                    cnt += int(graph[nx][ny])
                    q.append((nx,ny))
        cnt_lst.append(cnt)
        return

    for i in range(len(graph)):
        for j in range(len(graph[0])):
            if graph[i][j] != 'X' and not visited[i][j]:
                visited[i][j] = True 
                bfs(i,j,0)
    
    if cnt_lst:
        cnt_lst.sort()
    else:
        cnt_lst.append(-1)
    return cnt_lst

File: 9주차/김용상/무인도여행.java
import java.util.*;
class Solution {
    public int[] solution(String[] maps) {
        int[] answer = {};
        int []dx = new int[]{-1,1,0,0};
        int []dy = new int[]{0,0,-1,1};
        int n = maps.length;
        int m = maps[0].length();
        int[][]mp = new int[n][m];
        for (int i=0;i<n;i++){
            for (int j=0;j<m;j++){
                mp[i][j] = (int)maps[i].charAt(j)-'0';
            }
        }
        List<Integer>lst = new ArrayList<>();
        boolean[][]visited = new boolean[n][m];
        for (int i=0;i<n;i++){
            for (int j=0;j<m;j++){                
                if(visited[i][j]||mp[i][j]>=10)
                    continue;
                int cnt = 0;
                Deque<Node>dq = new ArrayDeque<>();
                dq.add(new Node(i,j,mp[i][j]));
                while(!dq.isEmpty()){
                    Node poll = dq.poll();
                    if (visited[poll.x][poll.y])
                        continue;
                    visited[poll.x][poll.y] = true; 
                    cnt+=poll.sm;
                    for(int k=0;k<4;k++){
                        int nx = dx[k]+poll.x;
                        int ny = dy[k]+poll.y;
                        if (nx>=0&&nx<n&&ny>=0&&ny<m&&!visited[nx][ny]&&mp[nx][ny]<10){
                            dq.add(new Node(nx,ny,mp[nx][ny]));
                        }
                    }            
                }
                if (cnt!=0)
                    lst.add(cnt);

            }
        }
        if (lst.size()==0) return new int[]{-1};
        Collections.sort(lst);                   
        answer = new int[lst.size()];
        for (int i=0;i<lst.size();i++){
            answer[i] = lst.get(i);
        }
        return answer;
    }
}
class Node{
    int x;
    int y;
    int sm;
    public Node(int x,int y,int sm){
        this.x = x;
        this.y = y;
        this.sm = sm;
    }
}
